cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp

```diff
@@ -150,56 +150,71 @@ sycl::event kernels_fp<Float>::select(sycl::queue& queue,
     std::int32_t* indices_ptr = indices.get_mutable_data();
     const auto fp_max = dal::detail::limits<Float>::max();
 
-    auto event = queue.submit([&](sycl::handler& cgh) {
-        cgh.depends_on(deps);
-        cgh.parallel_for<select_min_distance<Float>>(
-            bk::make_multiple_nd_range_2d({ wg_size, row_count }, { wg_size, 1 }),
-            [=](sycl::nd_item<2> item) {
-                auto sg = item.get_sub_group();
-                const std::int64_t sg_id = sg.get_group_id()[0];
-                const std::int64_t wg_id = item.get_global_id(1);
-                const std::int64_t sg_num = sg.get_group_range()[0];
-                const std::int64_t sg_global_id = wg_id * sg_num + sg_id;
-                if (sg_global_id >= row_count)
-                    return;
-                const std::int64_t in_offset = sg_global_id * stride;
-                const std::int64_t out_offset = sg_global_id;
+    const std::int64_t num_rows =
+        std::min(static_cast<std::int64_t>(bk::device_max_mem_alloc_size(queue) / sizeof(Float)),
+                 static_cast<std::int64_t>(std::numeric_limits<int>::max() / wg_size));
+    std::vector<sycl::event> events;
+    for (std::int64_t start_row = 0; start_row < row_count; start_row += num_rows) {
```

Comment:
```suggestion
const auto block_size = propose_block_size<Float>(queue, row_count);
const bk::uniform_blocking blocking(row_count, block_size);

std::vector<sycl::event> events(blocking.get_block_count());
for (std::int64_t b = 0; b < blocking.get_block_count(); ++b) {

    const auto first_row = blocking.get_block_start_index(b);
    const auto last_row = blocking.get_block_end_index(b);
```

Reply:
Please use well established functionality. Also I would recommend not to create so many events